File: macos/Sources/Voicekey/Core/SoundFX.swift
//
//  SoundFX.swift
//  録音開始/停止の操作音（コード合成・音源ファイル不要）
//
//  なぜコード合成か: 音源ファイルをバンドルに増やさず、開始=上昇・停止=下降の
//  短いブリップを正弦波で生成する。マイク録音の AVAudioEngine とは完全に独立した
//  再生専用エンジンを持ち、再生は fire-and-forget（音声入力パイプラインに待ちを足さない）。
//  失敗（オーディオ経路の一時不整合等）は握りつぶす＝音が鳴らないだけで実害はない。
//

import AVFoundation
import Foundation

/// 操作音の再生。`SoundFX.shared.play(.start / .stop)` を呼ぶだけ（撃ちっぱなし）。
final class SoundFX {

    /// アプリ全体で 1 つの再生エンジンを使い回す（毎回作ると起動コストがかかる）
    static let shared = SoundFX()

    /// 鳴らす操作音の種類
    enum Kind {
        /// 録音開始（2 音上昇のブリップ）
        case start
        /// 録音停止（下降のブリップ）
        case stop
    }

    /// 再生用の合成音を作る/鳴らすシリアルキュー。呼び出しスレッドをブロックしないため、
    /// バッファ生成・スケジューリングをすべてこのキューへ逃がす（クリティカルパス外）。
    private let queue = DispatchQueue(label: "com.voicekey.soundfx", qos: .userInitiated)

    private let engine = AVAudioEngine()
    private let player = AVAudioPlayerNode()
    /// 合成に使うフォーマット（モノラル 44.1kHz float）
    private let format = AVAudioFormat(standardFormatWithSampleRate: 44100, channels: 1)!
    private var isConfigured = false

    private init() {}

    /// 操作音を鳴らす（fire-and-forget）。呼び出し側は即座に戻る。
    func play(_ kind: Kind) {
        queue.async { [self] in
            guard let buffer = makeBuffer(for: kind) else { return }
            configureIfNeeded()
            // ルート変更等でエンジンが止まっていたら起動し直す（失敗は無視）
            if !engine.isRunning {
                try? engine.start()
            }
            guard engine.isRunning else { return }
            player.scheduleBuffer(buffer, at: nil, options: [], completionHandler: nil)
            if !player.isPlaying { player.play() }
        }
    }

    /// エンジンのノード接続を一度だけ行う
    private func configureIfNeeded() {
        guard !isConfigured else { return }
        engine.attach(player)
        engine.connect(player, to: engine.mainMixerNode, format: format)
        isConfigured = true
    }

    /// 種類に応じた短い正弦波ブリップを合成する。
    /// 開始=2 音上昇（低→高）、停止=下降（高→低）。両端をフェードしてクリック音を防ぐ。
    private func makeBuffer(for kind: Kind) -> AVAudioPCMBuffer? {
        let sampleRate = format.sampleRate
        // 各トーンの長さ（合計 100〜150ms に収める）
        let toneDuration = 0.065
        // 開始は 2 音（上昇）、停止は 2 音（下降）
        let freqs: [Double]
        switch kind {
        case .start: freqs = [660, 990]   // 上昇（E5→B5 相当）
        case .stop: freqs = [880, 587]    // 下降（A5→D5 相当）
        }

        let framesPerTone = Int(sampleRate * toneDuration)
        let totalFrames = framesPerTone * freqs.count
        guard let buffer = AVAudioPCMBuffer(
            pcmFormat: format, frameCapacity: AVAudioFrameCount(totalFrames)
        ) else { return nil }
        buffer.frameLength = AVAudioFrameCount(totalFrames)
        guard let samples = buffer.floatChannelData?[0] else { return nil }

        let amplitude: Float = 0.1   // 控えめな音量（程よく小さく）
        let twoPi = 2.0 * Double.pi
        // 前後 6ms を線形フェードしてプチッというクリックを消す
        let fadeFrames = Double(min(framesPerTone, Int(sampleRate * 0.006)))
        for (toneIndex, freq) in freqs.enumerated() {
            let base = toneIndex * framesPerTone
            for i in 0..<framesPerTone {
                let t = Double(i) / sampleRate
                let pos = Double(i)
                var env = 1.0
                if pos < fadeFrames { env = pos / fadeFrames }
                let tail = Double(framesPerTone) - pos
                if tail < fadeFrames { env = min(env, tail / fadeFrames) }
                samples[base + i] = Float(sin(twoPi * freq * t) * env) * amplitude
            }
        }
        return buffer
    }
}

File: src/core/sound_fx.py
"""
操作音モジュール（Windows・録音開始/停止の効果音・音源ファイル不要）

Mac 版 SoundFX.swift と同じく、開始=2 音上昇・停止=2 音下降の短いブリップを正弦波で
その場合成する（バンドルに音源ファイルを増やさない）。Mac は AVAudioEngine の再生専用
エンジン、Windows は既存依存の sounddevice で出力する。いずれもマイク録音とは独立した
再生で、fire-and-forget（音声入力パイプラインに待ちを足さない）。失敗（オーディオ経路の
一時不整合等）は握りつぶす＝音が鳴らないだけで実害はない。

テスト容易性: 波形合成は純粋関数 `make_buffer(kind)`（numpy のみ・音声ハードウェア不要）に
分離し、実際の再生（sounddevice）は `_default_player` に隔離して差し替え可能にしてある。
Mac と定数（周波数・長さ・振幅・フェード）を一致させること。
"""

import threading
from typing import Callable, Optional

import numpy as np

from ..utils.logger import get_logger

logger = get_logger(__name__)

# 合成・再生のパラメータ（Mac 版 SoundFX と一致させる）
SAMPLE_RATE = 44100      # 合成フォーマット（モノラル 44.1kHz float）
TONE_DURATION = 0.065    # 各トーンの長さ（秒）
AMPLITUDE = 0.1          # 控えめな音量
FADE_SEC = 0.006         # 前後 6ms を線形フェードしてクリック音を消す

# 開始=2 音上昇（E5→B5 相当）、停止=2 音下降（A5→D5 相当）
_FREQS = {
    "start": (660.0, 990.0),
    "stop": (880.0, 587.0),
}


def make_buffer(kind: str) -> Optional[np.ndarray]:
    """種類に応じた短い正弦波ブリップを合成する（float32 モノラル）。

    Args:
        kind: "start"（開始・上昇）または "stop"（停止・下降）

    Returns:
        合成した波形（float32 1 次元配列）。未知の種類なら None。
    """
    freqs = _FREQS.get(kind)
    if freqs is None:
        return None

    frames_per_tone = int(SAMPLE_RATE * TONE_DURATION)
    fade_frames = min(frames_per_tone, int(SAMPLE_RATE * FADE_SEC))
    two_pi = 2.0 * np.pi

    # 前後 fade_frames を線形フェードする包絡線（1 トーンぶん）
    env = np.ones(frames_per_tone, dtype=np.float64)
    if fade_frames > 0:
        ramp = np.arange(fade_frames, dtype=np.float64) / fade_frames
        env[:fade_frames] = np.minimum(env[:fade_frames], ramp)
        # 末尾はミラーで落とす（残り frames が fade_frames を割ると env が小さくなる）
        env[-fade_frames:] = np.minimum(env[-fade_frames:], ramp[::-1])

    t = np.arange(frames_per_tone, dtype=np.float64) / SAMPLE_RATE
    tones = []
    for freq in freqs:
        tones.append(np.sin(two_pi * freq * t) * env * AMPLITUDE)
    return np.concatenate(tones).astype(np.float32)


def _default_player(buffer: np.ndarray) -> None:
    """既定の再生（sounddevice・マイク録音とは独立した出力ストリーム）。"""
    import sounddevice as sd

    sd.play(buffer, SAMPLE_RATE)


def _play_impl(kind: str, player: Optional[Callable[[np.ndarray], None]] = None) -> None:
    """合成 → 再生の本体（失敗は握りつぶす）。テストは player を注入して検証する。"""
    try:
        buffer = make_buffer(kind)
        if buffer is None:
            return
        (player or _default_player)(buffer)
    except Exception as e:  # 音が鳴らないだけで実害はないため握りつぶす
        logger.debug(f"操作音の再生に失敗（無視）: {e}")


def play(kind: str) -> None:
    """操作音を鳴らす（fire-and-forget）。呼び出し側は即座に戻る。

    Args:
        kind: "start" または "stop"
    """
    try:
        threading.Thread(target=_play_impl, args=(kind,), daemon=True).start()
    except Exception as e:
        logger.debug(f"操作音の投入に失敗（無視）: {e}")
